modules/core/src/test/scala/org/tessellation/infrastructure/rewards/RewardsSuite.scala

```diff
@@ -2,141 +2,109 @@ package org.tessellation.infrastructure.rewards
 import org.tessellation.dag.snapshot.epoch.EpochProgress
 import org.tessellation.ext.kryo._
+import org.tessellation.keytool.KeyPairGenerator
 import org.tessellation.kryo.KryoSerializer
 import org.tessellation.schema.ID.Id
-import org.tessellation.schema.balance.Amount
+import org.tessellation.schema.generators.chooseNumRefined
 import org.tessellation.sdk.sdkKryoRegistrar
 import org.tessellation.security.SecurityProvider
-import org.tessellation.security.hash.Hash
-import org.tessellation.security.hex.Hex
+import org.tessellation.security.key.ops.PublicKeyOps
 
 import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.NonNegLong
+import org.scalacheck.Gen
 import weaver.MutableIOSuite
 import weaver.scalacheck.Checkers
 
 object RewardsSuite extends MutableIOSuite with Checkers {
-  type Res = (KryoSerializer[IO], SecurityProvider[IO], Random[IO])
+  type GenIdFn = () => Id
+  type Res = (KryoSerializer[IO], SecurityProvider[IO], Random[IO], GenIdFn)
 
   override def sharedResource: Resource[IO, Res] =
     (
       KryoSerializer.forAsync[IO](dagSharedKryoRegistrar.union(sdkKryoRegistrar)),
       SecurityProvider.forAsync[IO],
       Resource.liftK(Random.scalaUtilRandom[IO])
-    ).mapN((_, _, _))
-
-  implicit val order: Ordering[Id] = (x: Id, y: Id) => 1
+    ).mapN {
+      case (ks, sp, r) =>
+        implicit val _sp: SecurityProvider[IO] = sp
+        (ks, sp, r, () => KeyPairGenerator.makeKeyPair.map(_.getPublic.toId).unsafeRunSync())
+    }
 
   val config = RewardsConfig.default
 
-  val totalPool = Amount(160000000000000000L)
-
-  val softStaking = SoftStaking.make(config.softStaking)
-  val dtm = DTM.make(config.dtm)
-  val stardust = StardustCollective.make(config.stardust)
-  val hash = Hash("")
-
-  val facilitators = NonEmptySet.fromSetUnsafe(
-    SortedSet(
-      Id(
-        Hex(
-          "f2b230bb6ca3db890e647c171adbaeea27fc0191b7d84790c020c7fe0e04003c30a62d475af1300d8205a2436d640a1a091b7d53feb5a11bbb000013f75bdcf9"
-        )
-      ),
-      Id(
-        Hex(
-          "c7f9a08bdea7ff5f51c8af16e223a1d751bac9c541125d9aef5658e9b7597aee8cba374119ebe83fb9edd8c0b4654af273f2d052e2d7dd5c6160b6d6c284a17c"
-        )
-      )
-    )
-  )
-
-  test("epoch 1 snapshot reward is valid") { res =>
-    implicit val (_, sp, rand) = res
-    val regular = Regular.make
-
-    for {
-      rewards <- Rewards.make[IO](config, softStaking, dtm, stardust, regular).pure[F]
-      txs <- rewards
-        .calculateRewards(
-          EpochProgress(1L),
-          facilitators
-        )
-      sum = txs.toList.map(_.amount.value.toLong).sum
+  val lowerBound: NonNegLong = EpochProgress.MinValue.value
+  val upperBound: NonNegLong = config.rewardsPerEpoch.keySet.max.value
+  val lowerBoundNoMinting: NonNegLong = NonNegLong.unsafeFrom(upperBound.value + 1)
+  val special: Seq[NonNegLong] =
+    config.rewardsPerEpoch.keys.flatMap(epochEnd => Seq(epochEnd.value, NonNegLong.unsafeFrom(epochEnd.value + 1L))).toSeq
```

Comment:
I check here the epoch's upper bound and also the next number after the upper bound to see if halving works correctly.

Reply:
I think (I'm not sure though) that the generator will put higher weights also on neighbors of special so +1/-1. But I guess it won't hurt to keep it.